# Importing libraries

In [3]:
import pandas as pd
import numpy as np
import os

In [5]:
# Assigning path
path = r'C:\Users\Shruthi\Desktop\CF_Tasks\21-06-2023_Instacart_Basket_Analysis'

In [6]:
ords_prods_merge = pd.read_pickle(os.path.join(path, '2. Data', 'Prepared data', 'orders_products_merged_new.pkl'))

In [7]:
# Creating subset
df = ords_prods_merge[:1000000]

In [8]:
df.shape

(1000000, 21)

In [9]:
df.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  ...  product_name aisle_id  department_id  prices  _merge  \
0          0  ...          Soda       77              7     9.0    both   
1          1  ...          Soda       77              7     9.0    both   
2          1  ...          Soda       77              7     9.0    both   
3          1  ...          Soda       77              7     9.0    both   
4          1  ...          Soda       77              7     9.0    both   

   price_range_loc    price_range_loc     busiest_day     busiest_days  \
0              NaN  Mid-range product  Regularly busy   Regularly busy   
1              NaN  Mid-range product  Regularly busy  Least busy days   
2              NaN  Mid-range product  Regularly busy  Least busy days   
3              NaN  Mid-range product      Least busy  Least busy days   
4              NaN  Mid-range product      Least busy  Least busy days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2           Most orders  
3        Average orders  
4           Most orders  

[5 rows x 21 columns]

In [10]:
# Grouping data
df.groupby('product_name')

In [11]:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [12]:
# Aggregation using mean() function
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [13]:
# Performing multiple aggregations
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

In [14]:
# Creating max_order column
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [15]:
# Checking
ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  ...  aisle_id department_id  prices  _merge   price_range_loc  \
0          0  ...        77             7     9.0    both               NaN   
1          1  ...        77             7     9.0    both               NaN   
2          1  ...        77             7     9.0    both               NaN   
3          1  ...        77             7     9.0    both               NaN   
4          1  ...        77             7     9.0    both               NaN   

     price_range_loc     busiest_day     busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy   Regularly busy        Average orders   
1  Mid-range product  Regularly busy  Least busy days        Average orders   
2  Mid-range product  Regularly busy  Least busy days           Most orders   
3  Mid-range product      Least busy  Least busy days        Average orders   
4  Mid-range product      Least busy  Least busy days           Most orders   

  max_order  
0        10  
1        10  
2        10  
3        10  
4        10  

[5 rows x 22 columns]

In [16]:
pd.options.display.max_rows = None

In [18]:
# Checking
ords_prods_merge.head(100)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
5    3367565        1    prior             6                   2   
6     550135        1    prior             7                   1   
7    3108588        1    prior             8                   1   
8    2295261        1    prior             9                   1   
9    2550362        1    prior            10                   4   
10   2968173       15    prior            15                   1   
11   1870022       15    prior            17                   2   
12   1911383       15    prior            18                   2   
13   2715276       15    prior            21                   1   
14    487368       15    prior            22                   1   
15   2293453       19    prior             2                   5   
16   1973799       19    prior             5                   6   
17    532817       19    prior             7                   4   
18   1573906       21    prior            10                   3   
19   1593000       31    prior            10                   3   
20   2231262       31    prior            17                   3   
21   2580647       43    prior             6                   4   
22   2187180       43    prior             9                   4   
23   2497897       52    prior             1                   1   
24   1318871       52    prior             2                   1   
25   1261384       52    prior             3                   1   
26    580568       52    prior             5                   2   
27   2428073       52    prior             7                   2   
28    342306       52    prior             9                   1   
29   1520353       52    prior            10                   2   
30   2542086       52    prior            11                   2   
31    944694       52    prior            12                   1   
32   1268191       52    prior            15                   2   
33   1122089       52    prior            17                   2   
34   1498922       52    prior            18                   2   
35    180919       52    prior            19                   3   
36   2695875       52    prior            21                   5   
37   3205960       67    prior             1                   3   
38    970922       67    prior             2                   1   
39    596585       67    prior             3                   1   
40   2037087       67    prior             4                   2   
41   2474111       67    prior             5                   3   
42   1568481       67    prior             6                   3   
43   2048527       67    prior             8                   2   
44   2389031       67    prior             9                   2   
45   2378562       67    prior            10                   1   
46   1515698       67    prior            11                   1   
47   1928050       67    prior            12                   2   
48   1949274       67    prior            13                   1   
49    551869       67    prior            14                   1   
50    768572       67    prior            16                   4   
51    150528       67    prior            18                   5   
52     29758       67    prior            19                   0   
53   1558976       67    prior            21                   1   
54    863974       67    prior            22                   1   
55   2308399       67    prior            23                   1   
56   2759128       81    prior             3                   1   
57   1812985       81    prior             5                

In [19]:
# Creating flag
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [20]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [21]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [22]:
ords_prods_merge['loyalty_flag'].value_counts(dropna= False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [23]:
# Checking
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

# Finding the mean orders

In [24]:
# Grouping by "product_name" 
ords_prods_merge.groupby('product_name')

In [25]:
# Split the data into groups by "department_id" and apply agg function
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

-- There were differences in the mean orders when calculated using the entire dataframe rather than the subset: the minimum mean order in the subset is 11.3 while it is 15.2 in the dataframe, and the maximum mean in the subset is 19.5, 20.1 in the dataframe (not including department 21, the "missing" category).

# Creating loyalty flag

In [26]:
# Creating max_order column
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [27]:
pd.options.display.max_rows = None

In [28]:
# Checking
ords_prods_merge.head(100)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
5    3367565        1    prior             6                   2   
6     550135        1    prior             7                   1   
7    3108588        1    prior             8                   1   
8    2295261        1    prior             9                   1   
9    2550362        1    prior            10                   4   
10   2968173       15    prior            15                   1   
11   1870022       15    prior            17                   2   
12   1911383       15    prior            18                   2   
13   2715276       15    prior            21                   1   
14    487368       15    prior            22                   1   
15   2293453       19    prior             2                   5   
16   1973799       19    prior             5                   6   
17    532817       19    prior             7                   4   
18   1573906       21    prior            10                   3   
19   1593000       31    prior            10                   3   
20   2231262       31    prior            17                   3   
21   2580647       43    prior             6                   4   
22   2187180       43    prior             9                   4   
23   2497897       52    prior             1                   1   
24   1318871       52    prior             2                   1   
25   1261384       52    prior             3                   1   
26    580568       52    prior             5                   2   
27   2428073       52    prior             7                   2   
28    342306       52    prior             9                   1   
29   1520353       52    prior            10                   2   
30   2542086       52    prior            11                   2   
31    944694       52    prior            12                   1   
32   1268191       52    prior            15                   2   
33   1122089       52    prior            17                   2   
34   1498922       52    prior            18                   2   
35    180919       52    prior            19                   3   
36   2695875       52    prior            21                   5   
37   3205960       67    prior             1                   3   
38    970922       67    prior             2                   1   
39    596585       67    prior             3                   1   
40   2037087       67    prior             4                   2   
41   2474111       67    prior             5                   3   
42   1568481       67    prior             6                   3   
43   2048527       67    prior             8                   2   
44   2389031       67    prior             9                   2   
45   2378562       67    prior            10                   1   
46   1515698       67    prior            11                   1   
47   1928050       67    prior            12                   2   
48   1949274       67    prior            13                   1   
49    551869       67    prior            14                   1   
50    768572       67    prior            16                   4   
51    150528       67    prior            18                   5   
52     29758       67    prior            19                   0   
53   1558976       67    prior            21                   1   
54    863974       67    prior            22                   1   
55   2308399       67    prior            23                   1   
56   2759128       81    prior             3                   1   
57   1812985       81    prior             5                

In [29]:
# Creating loyalty_flag
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [30]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [31]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [32]:
# Checking frequencies of loyalty flag
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

# Finding the mean of the product prices of loyal customers

In [33]:
# Grouping by "prices" 
ords_prods_merge.groupby('prices')

In [34]:
# Split the data into groups by "loyalty_flag" and calculate the mean for each "prices" column
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
Loyal customer    10.386336
New customer      13.294670
Regular customer  12.495717

# Creating spending flag

In [35]:
# Creating column mean_price and grouping by user id and applying mean function on prices column
ords_prods_merge['mean_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [36]:
# Checking
ords_prods_merge.head(100)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
5    3367565        1    prior             6                   2   
6     550135        1    prior             7                   1   
7    3108588        1    prior             8                   1   
8    2295261        1    prior             9                   1   
9    2550362        1    prior            10                   4   
10   2968173       15    prior            15                   1   
11   1870022       15    prior            17                   2   
12   1911383       15    prior            18                   2   
13   2715276       15    prior            21                   1   
14    487368       15    prior            22                   1   
15   2293453       19    prior             2                   5   
16   1973799       19    prior             5                   6   
17    532817       19    prior             7                   4   
18   1573906       21    prior            10                   3   
19   1593000       31    prior            10                   3   
20   2231262       31    prior            17                   3   
21   2580647       43    prior             6                   4   
22   2187180       43    prior             9                   4   
23   2497897       52    prior             1                   1   
24   1318871       52    prior             2                   1   
25   1261384       52    prior             3                   1   
26    580568       52    prior             5                   2   
27   2428073       52    prior             7                   2   
28    342306       52    prior             9                   1   
29   1520353       52    prior            10                   2   
30   2542086       52    prior            11                   2   
31    944694       52    prior            12                   1   
32   1268191       52    prior            15                   2   
33   1122089       52    prior            17                   2   
34   1498922       52    prior            18                   2   
35    180919       52    prior            19                   3   
36   2695875       52    prior            21                   5   
37   3205960       67    prior             1                   3   
38    970922       67    prior             2                   1   
39    596585       67    prior             3                   1   
40   2037087       67    prior             4                   2   
41   2474111       67    prior             5                   3   
42   1568481       67    prior             6                   3   
43   2048527       67    prior             8                   2   
44   2389031       67    prior             9                   2   
45   2378562       67    prior            10                   1   
46   1515698       67    prior            11                   1   
47   1928050       67    prior            12                   2   
48   1949274       67    prior            13                   1   
49    551869       67    prior            14                   1   
50    768572       67    prior            16                   4   
51    150528       67    prior            18                   5   
52     29758       67    prior            19                   0   
53   1558976       67    prior            21                   1   
54    863974       67    prior            22                   1   
55   2308399       67    prior            23                   1   
56   2759128       81    prior             3                   1   
57   1812985       81    prior             5                

In [37]:
# Creating flags 
ords_prods_merge.loc[ords_prods_merge['mean_price'] < 10, 'spending_flag'] = 'Low spender'

In [38]:
ords_prods_merge.loc[ords_prods_merge['mean_price'] >= 10, 'spending_flag'] = 'High spender'

In [39]:
# Checking frequencies of spending flags 
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

# Creating order_frequency flag

In [41]:
# Creating new column median_days
ords_prods_merge['median_days'] = ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)

In [42]:
# Checking
ords_prods_merge.head(100)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
5    3367565        1    prior             6                   2   
6     550135        1    prior             7                   1   
7    3108588        1    prior             8                   1   
8    2295261        1    prior             9                   1   
9    2550362        1    prior            10                   4   
10   2968173       15    prior            15                   1   
11   1870022       15    prior            17                   2   
12   1911383       15    prior            18                   2   
13   2715276       15    prior            21                   1   
14    487368       15    prior            22                   1   
15   2293453       19    prior             2                   5   
16   1973799       19    prior             5                   6   
17    532817       19    prior             7                   4   
18   1573906       21    prior            10                   3   
19   1593000       31    prior            10                   3   
20   2231262       31    prior            17                   3   
21   2580647       43    prior             6                   4   
22   2187180       43    prior             9                   4   
23   2497897       52    prior             1                   1   
24   1318871       52    prior             2                   1   
25   1261384       52    prior             3                   1   
26    580568       52    prior             5                   2   
27   2428073       52    prior             7                   2   
28    342306       52    prior             9                   1   
29   1520353       52    prior            10                   2   
30   2542086       52    prior            11                   2   
31    944694       52    prior            12                   1   
32   1268191       52    prior            15                   2   
33   1122089       52    prior            17                   2   
34   1498922       52    prior            18                   2   
35    180919       52    prior            19                   3   
36   2695875       52    prior            21                   5   
37   3205960       67    prior             1                   3   
38    970922       67    prior             2                   1   
39    596585       67    prior             3                   1   
40   2037087       67    prior             4                   2   
41   2474111       67    prior             5                   3   
42   1568481       67    prior             6                   3   
43   2048527       67    prior             8                   2   
44   2389031       67    prior             9                   2   
45   2378562       67    prior            10                   1   
46   1515698       67    prior            11                   1   
47   1928050       67    prior            12                   2   
48   1949274       67    prior            13                   1   
49    551869       67    prior            14                   1   
50    768572       67    prior            16                   4   
51    150528       67    prior            18                   5   
52     29758       67    prior            19                   0   
53   1558976       67    prior            21                   1   
54    863974       67    prior            22                   1   
55   2308399       67    prior            23                   1   
56   2759128       81    prior             3                   1   
57   1812985       81    prior             5                

In [43]:
# Creating flags 
ords_prods_merge.loc[ords_prods_merge['median_days'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [44]:
ords_prods_merge.loc[(ords_prods_merge['median_days'] > 10) & (ords_prods_merge['median_days'] <= 20), 'frequency_flag'] = 'Regular customer'

In [45]:
ords_prods_merge.loc[ords_prods_merge['median_days'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [46]:
# Checking frequencies of frequency flag
ords_prods_merge['frequency_flag'].value_counts(dropna=False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: frequency_flag, dtype: int64

In [47]:
# Exporting df to pickle file
ords_prods_merge.to_pickle(os.path.join(path, '2. Data','Prepared data', 'orders_products_aggregated.pkl'))